synapse/lib/version.py

```diff
@@ -2,6 +2,182 @@
+
+vseps = ('.', '-', '_', '+')
+mask32 = 0xFFFFFFFF
+mask96 = 0xFFFFFFFFFFFFFFFFFFFFFFFF
+semver_re = r'''^(?P<maj>(0(?![0-9])|[1-9][0-9]*))\.(?P<min>(0(?![0-9])|[1-9][0-9]*))\.(?P<pat>(0(?![0-9])|[1-9][0-9]*))(\-(?P<pre>([0-9A-Za-z\-\.]+)))?(\+(?P<bld>([0-9A-Za-z\.\-]+)))?$'''
+
+def parseSemver(text):
+    '''
+    Parse a Semantic Version string into is component parts.
+
+    Args:
+        text (str): A text string to parse into semver components. This string has whitespace and leading 'v'
+        characters stripped off of it.
+
+    Examples:
+        Parse a string into it semvar parts::
+
+            parts = parseSemver('v1.2.3')
+
+    Returns:
+        dict: The dictionary will either be empty, contain the keys 'major', 'minor' and 'patch' pointing to integer
+        values.  The dictionary may also contain keys for 'build' and 'prerelease' information if that data is parsed
+        out of a semver string.  An empty dictionary indicates that the input string was not a Semantic Version
+        compliant string.
+    '''
+    # eat whitespace and leading chars common on version strings
+    txt = text.strip().lstrip('vV')
+    ret = {}
+
+    m = re.match(semver_re, txt)
+    if not m:
+        return {}
+    d = m.groupdict()
+    ret['major'] = int(d.get('maj'))
+    ret['minor'] = int(d.get('min'))
+    ret['patch'] = int(d.get('pat'))
+
+    pre = d.get('pre')
+    bld = d.get('bld')
+
+    if pre:
+        # Validate pre
+        parts = pre.split('.')
+        for part in parts:
+            if not part:
+                return {}
+            try:
+                int(part)
+            except ValueError:
+                continue
+            else:
+                if part[0] == '0' and len(part) > 1:
+                    return {}
+        ret['prerelease'] = pre
+
+    if bld:
+        # Validate bld
+        parts = bld.split('.')
+        for part in parts:
+            if not part:
+                return {}
+        ret['build'] = bld
+
+    return ret
+
+def packVersion(major, minor=0, patch=0):
+    '''
+    Pack a set of major/minor/patch integers into a single integer for storage.
+
+    Args:
+        major (int): Major version level integer.
+        minor (int): Minor version level integer.
+        patch (int): Patch version level integer.
+
+    Returns:
+        int:  System normalized integer value to represent a software version.
+    '''
+
+    ret = patch & mask32
+    ret = ret | (minor & mask32) << 32
+    ret = ret | (major & mask32) << 32 * 2
+    return ret
+
+def unpackVersion(ver):
+    '''
+    Unpack a system normalized integer representing a softare version into its component parts.
+
+    Args:
+        ver (int): System normalized integer value to unpack into a tuple.
+
+    Returns:
+        (int, int, int): A tuple containing the major, minor and patch values shifted out of the integer.
+    '''
+    major = (ver & mask96) >> 32 * 2
+    minor = (ver << 32 & mask96) >> 32 * 2
```

Comment:
i think these can be more optimally calculated like this:
```
major = (ver >> 64) & mask32
minor = (ver >> 32) & mask32
patch = ver & mask32
```

Reply:
Changing this to the following keeps the 96 bit constraint in place and makes it more readable:
```
    ver = ver & mask96
    major = (ver >> 64) & mask32
    minor = (ver >> 32) & mask32
    patch = ver & mask32
    return major, minor, patch
```

Reply:
done

Reply:
the iniitial mask isn't required